pgml-dashboard/app/middleware.py

```diff
@@ -45,7 +46,27 @@ def __call__(self, request):
                 token = auth_header.split(" ")[-1]
 
             token = jwt.decode(token, settings.SECRET_KEY, algorithms=["HS256"])
+            request.dashboard_uuid = token["dashboard"]
+            if "shard" in token:
+                request.shard = token["shard"]
         except Exception:
             return HttpResponse(PERMISSION_DENIED, status=403)
 
         return self.get_response(request)
+
+
+class MultiTenantMiddleware:
+    """Handle dashboard using multiple PostgresML instances.
+
+    PostgresML instances must be proxied by PgCat."""
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        if hasattr(request, "shard"):
+            with connection.cursor() as cursor:
+                shard = request.shard
+                # https://github.com/levkk/pgcat
+                cursor.execute(f"SET SHARD TO '{shard}'")
```

Comment:
What pooling mode is pgbouncer set to? I'm guessing this requires a transaction?